File: CrackingCodingInterview/Ch. 4/Ch. 4 -1/Ch. 4 -1/Node.cpp

#ifndef NODE_CPP
#define NODE_CPP


#include "Node.h"


template <class T>
Node<T>::Node()
{
	data = NULL;
	children = nullptr;
}

template<class T>
Node<T>::Node(T newData)
{
	data = newData;
	children = nullptr;
}



template <class T>
Node<T>::~Node()
{
	std::cout << "node destructor" << std::endl;
	delete children;
}

template<class T>
void Node<T>::addNeighbor( Node<T>* const newNeighbor )
{
	if (!this->children) {
		this->children = new std::vector<Node<T>*>;
	}


		this->children->push_back(newNeighbor);
}



template<class T>
std::string Node<T>::printNeighbor()
{
	std::string result;
	if (children) {
		

		for (int i = 0; i < children->size(); i++) {
			result += std::to_string(children->at(i)->getData()) + " ";
		}
	}
		
	
	return result;
}

template<class T>
T  const & Node<T>:: getData () const
{
	return T(data);
}

template<class T>
void Node<T>::setData(T newData)
{
	this->data = newData;
}

template<class T>
std::vector<Node<T>*>* const & Node<T>::getChildren() const
{
	return this->children;
}

template<class T>
void Node<T>::setChildren(std::vector<Node<T>*>* newChildren)
{
	this->children = newChildren;
}

#endif // !NODE_CPP

File: Java Practice/Data Structure/Graph/src/Node.java
import java.util.ArrayList;

public class Node<T>{
	private T data;
	private ArrayList<T> neighbors;
	private boolean visited;
	public Node(T object) {
		data = object;
		visited = false;
		neighbors = new ArrayList<T>();
	}
			
	public Node() {
		data = null;
	}
	public T getData() {
		return data;
	}
	
	public boolean hasVisited() {
		return visited;
	}
	
	public void visitied() {
		visited = true;
	}
	
	public boolean addNeighbor(T newNeighbor) {
		
		for(T neighborIterator: neighbors ) {
			if(neighborIterator == newNeighbor) {
				return false;
			}
		}
		this.neighbors.add(newNeighbor);
		return true;
	}
	
	public ArrayList<T> getNeighbors(){ 
		return neighbors;
	}
	
	public void printNeighbor() {
		for(T neighborIterator : neighbors) {
			System.out.print( " -> " + neighborIterator);
		}
		
		System.out.println("");
	}
	
	
	
}
